### PR TITLE
Rebuild for libboost 1.86 with matching patch

### LINKED ISSUE #235: Boost 1.86 needs a patch to work
Hello everyone,

I am trying to increase the maintenance of legacy FEniCS here on conda-forge. I have noticed that the rebuild for boost 1.86 at #234  does not work currently due to changes in the boost API https://github.com/conda-forge/fenics-feedstock/pull/234

At the moment, the failure seems to be related to the filesystem API, where some functions got deprecated. 

I would be happy to help out and fix this issue, but I am not so familiar with conda-forge. To me, it seems like a patch is needed that has to be included in the recipe.yml. If anyone could help and point me into the right direction how to do this, I am happy to create the patch and restore compatibility with boost.

@minrk Maybe you could help to get me started? It seems like you do a lot of work for FEniCS (and other projects) on conda-forge.

Thanks a lot.

### PR DESCRIPTION
**Pull request by sblauth**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

upstream patch: https://bitbucket.org/fenics-project/dolfin/pull-requests/577

closes #234
closes #235

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -15,9 +15,10 @@ source:
       - hdf5-1.12.patch
       - fix-xdmf.patch
       - python-cmake-args.patch
+      - boost-1.86.patch
```
Comment (minrk) [inline recipe/meta.yaml:18]: feel free to add yourself to the recipe-maintainers down at the bottom of this file

**Comment by sblauth** [comment]
There seems to be an issue with the linux builds and cmake. The error is

```
make[2]: *** No rule to make target '/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/libz.so', needed by '/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/work/python/build/lib.linux-x86_64-cpython-311/dolfin/cpp.cpython-311-x86_64-linux-gnu.so'.  Stop.
  make[1]: *** [CMakeFiles/Makefile2:87: CMakeFiles/cpp.dir/all] Error 2
  make: *** [Makefile:91: all] Error 2
  Traceback (most recent call last):
    File "<string>", line 2, in <module>
    File "<pip-setuptools-caller>", line 34, in <module>
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/work/python/setup.py", line 76, in <module>
      setup(name='fenics-dolfin',
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/__init__.py", line 117, in setup
      return distutils.core.setup(**attrs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/_distutils/core.py", line 183, in setup
      return run_commands(dist)
             ^^^^^^^^^^^^^^^^^^
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/_distutils/core.py", line 199, in run_commands
      dist.run_commands()
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/_distutils/dist.py", line 954, in run_commands
      self.run_command(cmd)
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/dist.py", line 995, in run_command
      super().run_command(command)
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/_distutils/dist.py", line 973, in run_command
      cmd_obj.run()
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/command/bdist_wheel.py", line 381, in run
      self.run_command("build")
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/_distutils/cmd.py", line 316, in run_command
      self.distribution.run_command(command)
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/dist.py", line 995, in run_command
      super().run_command(command)
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/_distutils/dist.py", line 973, in run_command
      cmd_obj.run()
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/_distutils/command/build.py", line 135, in run
      self.run_command(cmd_name)
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/_distutils/cmd.py", line 316, in run_command
      self.distribution.run_command(command)
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/dist.py", line 995, in run_command
      super().run_command(command)
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/site-packages/setuptools/_distutils/dist.py", line 973, in run_command
      cmd_obj.run()
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/work/python/setup.py", line 48, in run
      self.build_extension(ext)
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/work/python/setup.py", line 73, in build_extension
      subprocess.check_call(['cmake', '--build', '.'] + build_args, cwd=self.build_temp, env=env)
    File "/home/conda/feedstock_root/build_artifacts/fenics-pkgs_1733131048376/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.11/subprocess.py", line 413, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--config', 'Release']' returned non-zero exit status 2.
  error: subprocess-exited-with-error
```

I am not really sure how to fix this...

**Comment by minrk** [comment]
try adding `zlib` to the host dependencies of `fenics-dolfin` (the second output)

File: recipe/meta.yaml (hunk 1 above)
Comment (sblauth) [reply recipe/meta.yaml:18]: Okay, I did so with the next commit.

**Comment by minrk** [comment]
That's really weird, especially the 3.11 success. We can try to compare envs. Honestly, I'd also be okay skipping openmpi on osx-* if it's proving problematic.

**Comment by sblauth** [comment]
I could not find any noticeable differences in the envs between the different openmpi builds - mainly just different python versions and correspondingly changed packages. 

However, I noticed that even though the 3.11 build passed, it raised an error message
```
2024-12-02T13:47:32.7461530Z [Mac-1733145351225.local:38186] shmem: mmap: an error occurred while determining whether or not /tmp/ompi.Mac-1733145351225.501/jf.0/212140032/sm_segment.Mac-1733145351225.501.ca50000.0 could be created.
```

This seems to be similar to these issues: https://github.com/open-mpi/ompi/issues/12273 and https://github.com/open-mpi/ompi/issues/12784

There, it is suggested to add 
```
OMPI_MCA_btl_sm_backing_directory=/tmp
```
as this issue seems to be related to paths to the `tmp` directory which are too long. Could we try something like this for building under macos?

**Comment by minrk** [comment]
You can add

```
export OMPI_MCA_btl_sm_backing_directory=/tmp
```

to the beginning of `test-dolfin.sh` to test if that solves it

**Comment by minrk** [comment]
I realize the first crash is actually happening [here](https://github.com/conda-forge/fenics-feedstock/blob/fac360f71d337572d32393ada5435e158f2fea3e/recipe/build-dolfin.sh#L22), which probably needs the same env, if the tmpdir fix is going to work.

**Comment by sblauth** [comment]
Yes, you're right. As setting the `TMPDIR` variable did break more builds, I am trying to set `export OMPI_MCA_btl_sm_backing_directory=/tmp` in `build-dolfin.sh` and `test-dolfin.sh`

**Comment by minrk** [comment]
I don't either. It does seem like a bug in openmpi, but I don't know why fenics seems to be the only package affected:

```
[Mac-1733217746890:36830] [ 0] 0   libsystem_platform.dylib            0x00007ff8088f05ed _sigtramp + 29
[Mac-1733217746890:36830] [ 1] 0   ???                                 0x000000000000000e 0x0 + 14
[Mac-1733217746890:36830] [ 2] 0   libopen-pal.80.dylib                0x000000010cc58f6b opal_argv_join + 75
[Mac-1733217746890:36830] [ 3] 0   libmpi.40.dylib                     0x000000010bcccf78 ompi_rte_init + 3448
[Mac-1733217746890:36830] [ 4] 0   libmpi.40.dylib                     0x000000010bcd01ee ompi_mpi_instance_init + 382
[Mac-1733217746890:36830] [ 5] 0   libmpi.40.dylib                     0x000000010bcc9b97 ompi_mpi_init + 215
[Mac-1733217746890:36830] [ 6] 0   libmpi.40.dylib                     0x000000010bd34c55 MPI_Init_thread + 117
```

the error being a segfault in `opal_argv_join` makes me think it might be related to something like https://github.com/open-mpi/ompi/issues/11826 .

We could try adding `mpiexec -n 1`  before invoking `python`. If that works, I don't think we should publish builds with openmpi, but instead skip openmpi builds on mac until this is fixed in openmpi itself.

**Comment by minrk** [comment]
fwiw, I can reproduce the warning locally on my mac with `python -c 'from mpi4py import MPI; print(MPI.COMM_WORLD.rank)'`, _even when TMPDIR is set_. It only goes away if I add `mpiexec -n 1`, which makes me think there's something wrong with mpi initialization in openmpi on mac.

**Comment by sblauth** [comment]
I have updated `build-dolfin.sh` and `test-dolfin.sh` to use `mpiexec -n 1 python ...` instead of `python ...`
I haven't changed this for the actual pytest invocation, so that we can also see if this fails even though the build succeeds.

**Comment by minrk** [comment]
After this resolves, I can also update the feedstock to cross-compile arm/ppc on linux, which should save a bunch of time by avoiding the use of emulated compilers. But I don't want to do too many things at once.

**Comment by sblauth** [comment]
Well, now the osx build proceed further. `build-dolfin.sh` finishes and now the error occurs at `test-dolfin.sh`, as you expected. Should I disable the openmpi builds for macos for the moment?

**Comment by minrk** [comment]
> Should I disable the openmpi builds for macos for the moment?

Yes, you can add `# [not osx]`  to the openmpi entry [here](https://github.com/conda-forge/fenics-feedstock/blob/fac360f71d337572d32393ada5435e158f2fea3e/recipe/conda_build_config.yaml#L3) and then re-render the recipe (you can request a bot rerender with [this comment](https://conda-forge.org/docs/maintainer/infrastructure/#conda-forge-admin-please-rerender) after you push).

**Comment by sblauth** [comment]
@conda-forge-admin, please rerender

**Comment by sblauth** [comment]
Great, it seems like the rerendering can't happen because I originally forked the bot's fork of the feedstock. Is there an easy way to fix this? Or do I have to fork the original feedstock and recreate this PR from there for a re-rendering to work?

**Comment by minrk** [comment]
I'm not sure, that's weird. After this PR, you might delete your fork and create a new one from this repo for future work.

In the meantime, https://github.com/sblauth/fenics-feedstock_bot/pull/1 is a manual rerender

**Comment by sblauth** [comment]
Thanks a lot. I will delete this fork after this PR and fork the original feedstock.

**Comment by minrk** [comment]
Thank you!

**Pull request merged**